src/helm/benchmark/runner.py

```diff
@@ -138,10 +143,24 @@ def run_one(self, run_spec: RunSpec):
         adapter: Adapter = AdapterFactory.get_adapter(run_spec.adapter_spec, self.tokenizer_service)
 
         instances: List[Instance]
-        if not self.skip_instances:
-            # Create the instances of the scenario
-            with htrack_block("scenario.get_instances"):
-                instances = scenario.get_instances()
+        if self.skip_instances:
+            instances = []
+        else:
+            if self.read_instances_from_file:
+                with open(os.path.join(scenario.output_path, "instances.json")) as f:
+                    json_instances: List[Dict[str, Any]] = json.load(f)
+                instances = [dacite.from_dict(Instance, instance) for instance in json_instances]
+            else:
+                # Create the instances of the scenario
+                with htrack_block("scenario.get_instances"):
+                    instances = scenario.get_instances()
```

Comment:
There are a few problems to this approach:

1. We can have multiple `ScenarioSpecs` for the same same scenario. e.g. running `mmlu:subject=anatomy,model=simple1/model` and `mmlu:subject=philosophy,model=simple1/model` together will cause the `mmlu/instances.json` to only contain the instances for philosophy, because the `subject=philosophy` `AdapterSpec` overwrites the `mmlu/instances.json` file after `subject=anatomy`. So the path here needs to contain the adapter spec so we get different instances for different `ScenarioSpecs`.
2. I think we only want to export the instances in the CodaLab bundle, not the source scenarios data, so it would be good to put the file outside of the `scenarios` folder (e.g. `benchmark_output/instances`). This also avoids bugs if the scenario author is unlucky enough to have their own `instances.json` that's part of the source data that has a name collision with our special file.
3. This currently runs once per run, rather than once per `ScenarioSpec`. e.g. if the runs are `mmlu:subject=anatomy,model=openai/davinci-002` and `mmlu:subject=anatomy,model=openai/davinci`, it will write the identical `instances.json` file out twice, whereas it only needs to do so once. Possible approaches:
    a. Check if `instances.json` already exists, if so, skip rewriting the file.
    b. Build a set of all the unique `ScenarioSpecs` that we need first, and do the writing outside of `run_one`.

Reply:
Yes, good point. I revised it now so that:
- If the user specifies that it wants input_instances files to be saved, then we check if the path already exists and, if it does, we don't generate the file
- The generated instances parent folder is `benchmark_outputs/scenario_instances` folder.
- The generated instances folder contains the name of the scenario and then its corresponding args. (Following the discussion in the Google doc, since we're saving _all_ instances -- and not the augmented instances or the sampled instances -- I don't think we need to use anything from the adapter spec in the cache path).

I'm still not totally satisfied with my new logic here; I feel it's a bit clunky.